File: ChessMaker/Scripts/Game/Distance.js
"use strict";

Distance.prototype.equals = function (other) {
    return this.number == other.number && this.reference == other.reference;
};

Distance.prototype.getValue = function (previousStep, maxDist) {
    switch (this.reference) {
        case Distance.RelativeTo.None:
            return this.equals(Distance.Any) ? 1 : this.number; // the minimum for "any" is 1, and the max will be maxDist (from getRange)
        case Distance.RelativeTo.Max:
            return maxDist + this.number;
        case Distance.RelativeTo.Prev:
            if (previousStep != null)
                return previousStep.distance + this.number;
            return this.number;
        default:
            throw "Unexpected Distance.RelativeTo value in Distance.parseValue: " + this.reference;
    }
}

Distance.prototype.getRange = function (dist2, previousStep, maxDist) {
    var outDist1 = this.getValue(previousStep, maxDist);
    var outDist2;

    if (this.equals(Distance.Any))
        outDist2 = maxDist;
    else if (dist2 == null)
        outDist2 = outDist1;
    else
        outDist2 = dist2.getValue(previousStep, maxDist);

    if (outDist2 > maxDist)
        outDist2 = maxDist;
    return [outDist1, outDist2];
}

File: ChessmakerCore/Distance.cpp
#include "Distance.h"
#include "MoveStep.h"

#define DISTANCE_ANY -1
Distance Distance::Any = Distance(None, DISTANCE_ANY);
Distance Distance::Zero = Distance(None, 0);


Distance::Distance(RelativeTo_t relativeTo, int number)
{
	this->relativeTo = relativeTo;
	this->number = number;
}


Distance::~Distance()
{
	
}


int Distance::GetValue(MoveStep *previousStep, int maxDist)
{
	switch (relativeTo)
	{
	case None:
		return this == &Any ? 1 : number; // the minimum for "any" is 1, and the max will be maxDist (from getRange)
	case Max:
		return maxDist + number;
	case Prev:
		if (previousStep != 0)
			return previousStep->distance + number;
		return number;
	default:
		ReportError("Distance has unexpected relative-to value: %i\n", (int)relativeTo);
		return 0;
	}
}


int Distance::GetMaxValue(Distance *dist2, MoveStep *previousStep, int maxDist)
{
	if (this == &Any)
		return maxDist;

	if (dist2 == 0)
		return GetValue(previousStep, maxDist);
	
	return dist2->GetValue(previousStep, maxDist);
}